File: sources/iOS/ios-communications/Sources/iOSCommunications/ble/api/model/advertisement/BlePolarHrAdvertisement.swift

import Foundation

public class BlePolarHrAdvertisement {
    var currentData = Data()
    
    public var batteryStatus : Bool {
        get {
            if(currentData.isEmpty) {
                return false
            } else {
                return (currentData[0] & 0x01) == 1
            }
        }
    }
    
    public var sensorContact : Bool {
        get {
            if(currentData.isEmpty) {
                return false
            } else {
                return ((currentData[0] & 0x02) >> 1) == 1
            }
        }
    }
    
    public var advFrameCounter: UInt8 {
        get {
            if(currentData.isEmpty) {
                return 0xFF
            } else {
                return (currentData[0] & 0x1C) >> 2
            }
        }
    }
    private var previousAdvFrameCounter: UInt8 = 0
    
    public var broadcastBit: UInt8 {
        get {
            if(currentData.isEmpty) {
                return 0
            } else {
                return (currentData[0] & 0x20) >> 5
            }
        }
    }
    
    public var sensorDataType: UInt8 {
        get {
            if(currentData.isEmpty) {
                return 0
            } else {
                return (currentData[0] & 0x40) >> 6
            }
        }
    }
    
    public var statusFlags: UInt8 {
        get {
            if(currentData.isEmpty) {
                return 0
            } else {
                return (currentData[0] & 0x80) >> 7
            }
        }
    }
    
    public var khzCode: UInt8 {
        get {
            if(currentData.isEmpty) {
                return 0
            } else {
                return currentData[1]
            }
        }
    }
    
    public var fastAverageHr: UInt8 {
        get {
            if(currentData.isEmpty) {
                return 0
            } else {
                return (currentData[2] & 0x000000FF)
            }
        }
    }
    
    public var slowAverageHr: UInt8 {
        get {
            if(currentData.isEmpty) {
                return 0
            } else {
                if( currentData.count == 4 ){
                    return (currentData[3] & 0x000000FF)
                } else {
                    return (currentData[2] & 0x000000FF)
                }
            }
        }
    }
    
    public var isHrDataUpdated: Bool {
        get { return previousAdvFrameCounter != advFrameCounter }
    }
    
    public var isPresent: Bool {
        get {
            return currentData.count != 0
        }
    }
    
    public var hrValueForDisplay: UInt8 {
        get {
            return slowAverageHr
        }
    }
    
    public func processPolarManufacturerData(_ data: Data) {
        previousAdvFrameCounter = advFrameCounter
        currentData = data
    }
    
    public func resetToDefault() {
        currentData = Data()
    }
    
    public func description() -> String {
        return "HR: \(hrValueForDisplay) UC: \(advFrameCounter)"
    }
}

File: sources/Android/android-communications/library/src/main/java/com/polar/androidcommunications/api/ble/model/advertisement/BlePolarHrAdvertisement.kt
package com.polar.androidcommunications.api.ble.model.advertisement

import androidx.annotation.VisibleForTesting

class BlePolarHrAdvertisement {
    private var currentData: ByteArray = byteArrayOf()

    val batteryStatus: Int
        get() {
            return if (currentData.isNotEmpty()) {
                currentData[0].toInt() and 0x01
            } else {
                0
            }
        }

    val sensorContact: Int
        get() {
            return if (currentData.isNotEmpty()) {
                currentData[0].toInt() and 0x02 shr 1
            } else {
                0
            }
        }

    @VisibleForTesting
    var advFrameCounter = -1
        private set
        get() {
            return if (currentData.isNotEmpty()) {
                currentData[0].toInt() and 0x1C shr 2
            } else {
                -1
            }
        }

    private var previousAdvFrameCounter = 0

    val broadcastBit: Int
        get() {
            return if (currentData.isNotEmpty()) {
                currentData[0].toInt() and 0x20 shr 5
            } else {
                0
            }
        }

    private val sensorDataType: Int
        get() {
            return if (currentData.isNotEmpty()) {
                currentData[0].toInt() and 0x40 shr 6
            } else {
                0
            }
        }

    val statusFlags: Int
        get() {
            return if (currentData.isNotEmpty()) {
                currentData[0].toInt() and 0x80 shr 7
            } else {
                0
            }
        }

    val khzCode: Int
        get() {
            return if (currentData.isNotEmpty()) {
                currentData[1].toInt() and 0x000000FF
            } else {
                0
            }
        }
    val fastAverageHr: Int
        get() {
            return if (currentData.isNotEmpty()) {
                currentData[2].toInt() and 0x000000FF
            } else {
                0
            }
        }
    val slowAverageHr: Int
        get() {
            return if (currentData.isNotEmpty() && currentData.size == 4) {
                currentData[3].toInt() and 0x000000FF
            } else if (currentData.isNotEmpty() && currentData.size == 3) {
                currentData[2].toInt() and 0x000000FF
            } else {
                0
            }
        }

    var hrForDisplay = 0
        private set
        get() = slowAverageHr
    val isPresent: Boolean
        get() = currentData.isNotEmpty()
    val isHrDataUpdated: Boolean
        get() = previousAdvFrameCounter != advFrameCounter

    fun processPolarManufacturerData(data: ByteArray) {
        previousAdvFrameCounter = advFrameCounter
        currentData = data
    }

    fun resetToDefault() {
        currentData = byteArrayOf()
        advFrameCounter = -1
    }
}